engine/swarm/secrets.md

```diff
@@ -0,0 +1,770 @@
+        can't be used to sign certificates.
+
+          ```none
+          [server]
+          authorityKeyIdentifier=keyid,issuer
+          basicConstraints = critical,CA:FALSE
+          extendedKeyUsage=serverAuth
+          keyUsage = critical, digitalSignature, keyEncipherment
+          subjectAltName = DNS:localhost, IP:127.0.0.1
+          subjectKeyIdentifier=hash
+          ```
+
+    8.  Sign the site certificate.
+
+          ```bash
+          $ openssl x509 -req -days 750 -in "site.csr" -sha256 \
+              -CA "root-ca.crt" -CAkey "root-ca.key"  -CAcreateserial \
+              -out "site.crt" -extfile "site.cnf" -extensions server
+
+          Signature ok
+          subject=/C=US/ST=CA/L=San Francisco/O=Docker/CN=localhost
+          Getting CA Private Key
+          ```
+
+    9.  The `csr` and `cnf` files are not needed by the Nginx service, but you
+        will need them if you want to generate a new site certificate. Protect
+        the `key` files.
+
+3.  Create three secrets, representing the key, the certificate, and the
+    `nginx.conf`. You can store any file as a secret as long as it is smaller
+    than 500 KB. This allows you to decouple the key, certificate, and
+    configuration from the services that will use them.
+
+    ```bash
+    $ docker secret create site.key -f site.key site.key
+
+    2hvoi9mnnaof7olr3z5g3g7fp
+
+    $ docker secret create -f site.crt site.crt
+
+    aya1dh363719pkiuoldpter4b
+
+    $ docker secret create -f nginx.conf nginx.conf
+
+    zoa5df26f7vpcoz42qf2csth8
+
+    $ docker secret ls
+
+    ID                          NAME                  CREATED             UPDATED
+    2hvoi9mnnaof7olr3z5g3g7fp   site.key       58 seconds ago      58 seconds ago
+    aya1dh363719pkiuoldpter4b   site.crt       24 seconds ago      24 seconds ago
+    zoa5df26f7vpcoz42qf2csth8   nginx.conf     11 seconds ago      11 seconds ago
+    ```
+    {: id="step3"}
+
+4.  Create a service that runs `nginx` and has access to the three secrets. the
+    last part of the `docker service create` command creates a symbolic link
+    from the location of the `nginx.conf` secret to `/etc/nginx.conf.d/`, where
+    Nginx looks for extra configuration files, before actually starting Nginx.
+    This ensures that you don't need to rebuild your image if you change the
+    Nginx configuration.
+
+    **Note**: Normally you would create a Dockerfile which copies the `nginx.conf`
+    into place, build the image, and run a container using your custom image.
+    This example does it all in one step.
+
+    ```bash
+    $ docker service create --name nginx \
+                            --secret site.key \
+                            --secret site.crt \
+                            --secret nginx.conf \
+                            --publish 3000:443 \
+                            nginx:latest \
+    sh -c "ln -s /run/secrets/nginx.conf /etc/nginx/conf.d/ nginx.conf && nginx -g 'daemon off;'"
```

Comment:
this should use `exec` so that nginx is PID 1 inside the container after the symlink was created;

```
sh -c "ln -s /run/secrets/nginx.conf /etc/nginx/conf.d/ nginx.conf && exec nginx -g 'daemon off;'"
```